xtext.cpp

```diff
@@ -0,0 +1,2157 @@
+                printf("%sMSHB g%d c%d(=t:%x) [%x %s]\tadvance=(%d,%d)", indent, i, glyph_info[i].cluster,
+                            m_text[glyph_info[i].cluster], glyph_info[i].codepoint, glyphname,
+                            glyph_pos[i].x_advance>>6, glyph_pos[i].y_advance>>6);
+                if (glyph_pos[i].x_offset || glyph_pos[i].y_offset)
+                    printf("\toffset=(%d,%d)", glyph_pos[i].x_offset>>6, glyph_pos[i].y_offset>>6);
+                printf("\n");
+            }
+            printf("%sMSHB ---\n", indent);
+        #endif
+
+        // We need to set widths and flags on our original text.
+        // hb_shape() has modified buffer to contain glyphs, and text
+        // and buffer may desync (because of clusters, ligatures...)
+        // in both directions in a same run.
+        // Also, a cluster must not be cut, so we want to set the same
+        // width to all our original text chars that are part of the
+        // same cluster (so 2nd+ chars in a cluster will get a 0-width,
+        // and, when splitting lines, will fit on a line with the
+        // cluster leading char).
+        // So run along our original text (chars, t), and try to follow
+        // harfbuzz buffer (glyphs, hg), putting the advance of all
+        // the glyphs that belong to the same cluster (hcl) on the
+        // first char that started that cluster (and 0-width on the
+        // followup chars).
+        // It looks like Harfbuzz makes a cluster of combined glyphs
+        // even when the font does not have any or all of the required
+        // glyphs:
+        // When meeting a not-found glyph (codepoint=0, name=".notdef"),
+        // we record the original starting t of that cluster, and
+        // keep processing (possibly other chars with .notdef glyphs,
+        // giving them the width of the 'tofu' char), until we meet a char
+        // with a found glyph. We then hold on on this one, while we go
+        // measureSegment() the previous segment of text (that got .notdef
+        // glyphs) with a fallback font, and update the wrong widths
+        // and flags.
+
+        int final_width = 0;
+        int cur_cluster = 0;
+        int hg = 0;  // index in glyph_info/glyph_pos
+        int hcl = 0; // cluster number of glyph at hg
+        int t_notdef_start = -1;
+        int t_notdef_end = -1;
+        int notdef_width = 0;
+        for ( int t = start; t < end; t++ ) {
+            #ifdef DEBUG_MEASURE_TEXT
+                printf("%sMSHB t%d (=%x) ", indent, t, m_text[t]);
+            #endif
+            // Grab all glyphs that do not belong to a cluster greater that our char position
+            int cur_width = 0; // current cluster width
+            while ( hg < glyph_count ) {
+                hcl = glyph_info[hg].cluster;
+                if ( hcl <= t ) { // glyph still part of a previous cluster
+                    int advance = 0;
```

Comment:
Sidebar: I'd keep that in a matching data type (hb_position_t, AFAIK, which points to an int32_t, FWIW).

Reply:
Well, thought about that (adding the non >>6 advances, even returning them to frontend and have it do the >>6 with the cumulative widths to not accumulate rounding errors).
But no: with my recursive calls to measureSegments, I need to have consistent stuff with individual segment full advance = the sum of the individual x_advance I return for each glyph. So, better to end up doing the >>6 (well, trunc(round()) for better looking results) here.